File: examples/cpp/cli-argsbarg.cpp
#!/usr/bin/env -S shebangsy cpp
#!requires: github:bdombro/cpp-argsbarg@v0.5.0:argsbarg::argsbarg

// Small CLI with cpp-argsbarg: ``hello`` greets ``world`` by default; optional
// ``--name`` / ``-n`` and ``--verbose`` / ``-v``; falls back to ``hello`` when the
// command is missing or unknown.
//
// Usage:
//   ./examples/cpp/cli-argsbarg.cpp hello
//   ./examples/cpp/cli-argsbarg.cpp hello --name Ada
//   ./examples/cpp/cli-argsbarg.cpp hello -n Ada -v
//
// Expected (one block per usage line above, in order):
//   hello world
//
//   hello Ada
//
//   verbose mode
//   hello Ada

#include <argsbarg/argsbarg.hpp>
#include <iostream>

using namespace argsbarg;

int main(int argc, const char* argv[]) {
    auto greet = [](Context& ctx) {
        const auto name = ctx.string_opt("name").value_or("world");
        if (ctx.flag("verbose")) {
            std::cout << "verbose mode\n";
        }
        std::cout << "hello " << name << '\n';
    };

    Application{"minimaldemo"}
        .description("Tiny demo.")
        .fallback("hello", FallbackMode::MissingOrUnknown)
        .command(Leaf{"hello", "Say hello."}
                     .handler(greet)
                     .option(Opt{"name", "Who to greet."}.string().short_alias('n'))
                     .option(Opt{"verbose", "Enable extra logging."}.short_alias('v')))
        .run(argc, argv);
}

File: examples/swift/cli-argsbarg.swift
#!/usr/bin/env -S shebangsy swift
#!requires: bdombro/swift-argsbarg@0.2.0:ArgsBarg

/*
Small CLI with swift-argsbarg: the ``hello`` subcommand greets ``world`` by default,
supports ``--name`` / ``-n`` and ``--verbose`` / ``-v``, and falls back to ``hello``
when the command is missing or unknown.

Usage:
	./examples/swift/cli-argsbarg.swift hello
	./examples/swift/cli-argsbarg.swift hello --name Ada
	./examples/swift/cli-argsbarg.swift hello -n Ada -v

Expected (one block per usage line above, in order):
	hello world

	hello Ada

	verbose mode
	hello Ada
*/
import ArgsBarg

cliRun(
    CliCommand(
        name: "cli-argsbarg.swift",
        description: "Tiny demo.",
        children: [
            CliCommand(
                name: "hello",
                description: "Say hello.",
                options: [
                    CliOption(
                        name: "name",
                        description: "Who to greet.",
                        kind: .string,
                        shortName: "n"
                    ),
                    CliOption(
                        name: "verbose",
                        description: "Enable extra logging.",
                        shortName: "v"
                    ),
                ],
                handler: { ctx in
                    let name = ctx.stringOpt("name") ?? "world"
                    if ctx.flag("verbose") { print("verbose mode") }
                    print("hello \(name)")
                }
            )
        ],
        fallbackCommand: "hello",
        fallbackMode: .missingOrUnknown
    ))
